k8s-operator/apis/v1alpha1/types_tsrecorder.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=rec
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=`.status.conditions[?(@.type == "RecorderReady")].reason`,description="Status of the deployed TSRecorder resources."
+
+type TSRecorder struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec describes the desired recorder instance.
+	Spec TSRecorderSpec `json:"spec"`
+
+	// TSRecorderStatus describes the status of the recorder. This is set
+	// and managed by the Tailscale operator.
+	// +optional
+	Status TSRecorderStatus `json:"status"`
+}
+
+// +kubebuilder:object:root=true
+
+type TSRecorderList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []TSRecorder `json:"items"`
+}
+
+type TSRecorderSpec struct {
+	// Tags that the Tailscale node will be tagged with. Defaults to [tag:k8s].
+	// If you specify custom tags here, make sure you also make the operator
+	// an owner of these tags.
+	// See  https://tailscale.com/kb/1236/kubernetes-operator/#setting-up-the-kubernetes-operator.
+	// Tags cannot be changed once a TSRecorder node has been created.
```

Comment:
We should be able to add a CEL rule to actually enforce that this is not changed
https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/

Reply:
Actually, seems like I didn't do this for the Connector, so perhaps it was impossible to enforce immutable arrays

Reply:
I gave this a bit of a go, but couldn't make it work. IIRC we discussed this or similar offline and it _could_ be made mutable because it looks like we have the API method we'd need, but it's just not currently supported, so I won't put too much effort into the CEL for now as it would be nice to support that.